feat(curriculum): add weather app lab to cert (#56246)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md

```js
assert.exists(document.querySelector("img#weather-icon"));
```

You should have an element with the id `main-temperature` for displaying the main temperature.

```js
assert.exists(document.querySelector("#main-temperature"));
```

You should have an element with the id `feels-like` for displaying what the temperature feels like.

```js
assert.exists(document.querySelector("#feels-like"));
```

You should have an element with the id `humidity` for displaying the amount of humidity in air.

```js
assert.exists(document.querySelector("#humidity"))
```

You should have an element with the id `wind` element for displaying the wind speed.

```js
assert.exists(document.querySelector("#wind"));
```

You should have an element with the id `wind-gust` element for displaying the wind gust.

```js
assert.exists(document.querySelector("#wind-gust"));
```

You should have an element with the id `weather-main` element for displaying the main weather type.

```js
assert.exists(document.querySelector("#weather-main"));
```

You should have an element with the id `location` element for displaying the current location.

```js
assert.exists(document.querySelector("#location"));
```

> You should have a `clickHandler` function.

Reviewer: Not mentioned before

```js
assert.isFunction(clickHandler);
```

> You should have a `checkPermissions` function.

Reviewer: Not mentioned before

```js
assert.isFunction(checkPermissions);
```

> You should have a `showWeather` function.

Reviewer: Not mentioned before

```js
assert.isFunction(showWeather);
```

You should have a `getWeather` function.

```js

> assert.isFunction(getWeather);

Reviewer: It seems that it doesn't work with async functions :(

        const body = await fetch(`https://weather-proxy.freecodecamp.rocks/api/city/london`).then(
            data => data.json()
        )

        await new Promise(resolve => setTimeout(resolve, 4000));

        // construct the object from the data from the API
        const extractedData = helper(body);

        // check that all things are in place
        for (const id in extractedData) {
            if (id === "weather-icon") {
                assert.equal(document.querySelector('#weather-icon').src, extractedData[id]);
            } else {
                assert.include(document.querySelector(`#${id}`).innerText.toLowerCase(), `${extractedData[id]}`.toLowerCase());
            }
        }
    } catch (err) {
        throw new Error(err);
    }
}
```

If there is an error, your `getWeather` function should log the error to the console.

```js
const testArr = [];
const temp1 = fetch;
const temp2 = console.log;
const temp3 = console.error;
async () => {
  try {
    console.log = obj => {testArr.push(obj.toString())};
    console.error = obj => {testArr.push(obj.toString())};
    fetch = source => {throw new Error("This is a test error");}
    await getWeather("chicago");

>     assert.deepEqual(testArr, ["Error: This is a test error"]);

Reviewer 2: ```suggestion
    assert.include(testArr[0], "This is a test error");
    assert.lengthOf(testArr, 1);
```

   - You should have an element with the id `wind` element for displaying the wind speed.

   - You should have an element with the id `wind-gust` element for displaying the wind gust.

   - You should have an element with the id `weather-main` element for displaying the main weather type.

   - You should have an element with the id `location` element for displaying the current location.

1. You should have an asynchronous function named `getWeather` that fetches the weather information from the `https://weather-proxy.freecodecamp.rocks/api/city/<CITY>` API and returns it.

1. The `getWeather` asynchronous function should accept a city as argument.

1. You should handle any errors that occur within the `getWeather` function and log them to the console.

1. You should have a `showWeather` function that accepts a city as parameter.

1. The `showWeather` function should call the `getWeather` function to retrieve the weather data for a certain city.

1. If the `getWeather` function had an error and returns a falsy value, the `showWeather` function should only show an alert that says `Something went wrong, please try again later`.

> 1. If the data from `getWeather` are usable, the `showWeather` function should display the weather data in the corresponding elements.

Reviewer 2: ```suggestion
1. If the data from `getWeather` are usable, the `showWeather` function should display the weather data in the corresponding elements. If a certain value from the API is `undefined`, you should write `N/A` in the corresponsing element.
```

       await new Promise(resolve => setTimeout(resolve, 4000));
    assert.exists(document.querySelector("#wind"));
}
```

You should have an element with the id `wind-gust` element for displaying the wind gust.

```js
async () => {
    document.querySelector('select').value = "new york";
    document.querySelector('#get-forecast').click();
       await new Promise(resolve => setTimeout(resolve, 4000));
    assert.exists(document.querySelector("#wind-gust"));
}
```

You should have an element with the id `weather-main` element for displaying the main weather type.

```js
async () => {
    document.querySelector('select').value = "new york";
    document.querySelector('#get-forecast').click();
       await new Promise(resolve => setTimeout(resolve, 4000));
    assert.exists(document.querySelector("#weather-main"));
}
```

You should have an element with the id `location` element for displaying the current location.

```js
async () => {
    document.querySelector('select').value = "new york";
    document.querySelector('#get-forecast').click();
       await new Promise(resolve => setTimeout(resolve, 4000));
       assert.exists(document.querySelector("#location"));
}
```

> You should have a `clickHandler` function.
> 
> ```js
> assert.isFunction(clickHandler);
> ```
> 

Reviewer 2: ```suggestion
```

---
id: 66f12a88741aeb16b9246c59
title: Build a Weather App
challengeType: 14 
dashedName: lab-weather-app
demoType: onClick
---

# --description--

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

> You will use a weather API, the data you need is accessible like in this JSON example:

Reviewer: ```suggestion
You will use a weather API. The output data has the following format:
```
